File: C++/Arrays&Hashing/product_array_except_self.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Product of Array Except Self
     * 
     * Approach:
     * - Calculate prefix products (all elements to left)
     * - Calculate suffix products (all elements to right)
     * - Result for each index is prefix[i] * suffix[i]
     * 
     * Time Complexity: O(n) - three passes through array
     * Space Complexity: O(n) - for prefix and suffix arrays
     */
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        
        // Initialize prefix and suffix arrays with 1s
        vector<int> prefix(n, 1);
        vector<int> suffix(n, 1);
        
        // Calculate prefix products
        for (int i = 0; i < n; i++) {
            if (i == 0) continue;
            prefix[i] = prefix[i-1] * nums[i-1];
        }
        
        // Calculate suffix products
        for (int i = n-1; i >= 0; i--) {
            if (i == n-1) continue;
            suffix[i] = suffix[i+1] * nums[i+1];
        }
        
        // Calculate final result
        vector<int> result(n);
        for (int i = 0; i < n; i++) {
            result[i] = prefix[i] * suffix[i];
        }
        
        return result;
    }
};

/*
 * Detailed C++ Syntax Explanation:
 * 
 * 1. Vector Declaration:
 *    vector<int> prefix(n, 1)   // Create vector of size n filled with 1s
 *    vector<int> result(n)      // Create vector of size n (default 0s)
 * 
 * 2. Size Access:
 *    nums.size()                // Get size of vector
 * 
 * 3. For Loops:
 *    for (int i = 0; i < n; i++)    // Forward iteration
 *    for (int i = n-1; i >= 0; i--) // Backward iteration
 * 
 * 4. Vector Access:
 *    nums[i]                    // Access element at index i
 *    prefix[i-1]               // Access previous element
 * 
 * 5. Continue Statement:
 *    if (i == 0) continue;     // Skip rest of loop for first element
 */

File: python/Arrays&Hashing/product_array_except_self.py
from typing import List

class Solution:
    def productExceptSelf(self, nums : List[int]) -> List[int]:
        """
        Approach:
        - Use two arrays (prefix and suffix) to store the product of all elements to the left and right of each index.
        - prefix[i] contains the product of all elements to the left of i.
        - suffix[i] contains the product of all elements to the right of i.
        - The result for each index is prefix[i] * suffix[i].
        - This avoids using division and handles zeros in the input.

        Time Complexity: O(n), where n is the length of nums (three passes through the array).
        Space Complexity: O(n), for the prefix, suffix, and result arrays. Can be optimized to O(1) extra space (excluding output).
        """
        prefix = [1] * len(nums)
        suffix = [1] * len(nums)

        for i, n in enumerate(nums):
            if i == 0:
                pass
            else:
                prefix[i] = prefix[i-1] * nums[i-1]

        for i in range(len(nums)-1, -1, -1):
            if i == len(nums)-1:
                pass
            else:
                suffix[i] = suffix[i+1] * nums[i+1]

        result = []
        for i in range(len(nums)):
            result.append(prefix[i] * suffix[i])
        return result

# -------------------
# Test Cases
# -------------------
def test_product_array_except_self():
    sol = Solution()
    # Test case 1: Example from Leetcode
    nums1 = [1,2,3,4]
    assert sol.productExceptSelf(nums1) == [24,12,8,6]
    # Test case 2: Contains zero
    nums2 = [0,1,2,3]
    assert sol.productExceptSelf(nums2) == [6,0,0,0]
    # Test case 3: All ones
    nums3 = [1,1,1,1]
    assert sol.productExceptSelf(nums3) == [1,1,1,1]
    # Test case 4: Single element
    nums4 = [42]
    assert sol.productExceptSelf(nums4) == [1]
    # Test case 5: Two elements
    nums5 = [2,3]
    assert sol.productExceptSelf(nums5) == [3,2]
    print("All test cases passed!")

if __name__ == "__main__":
    test_product_array_except_self()
